tests/Unit/SentryClientCest.php

```diff
@@ -0,0 +1,548 @@
+
+        $I->assertEquals($expected, $body["contexts"]["browser"], "Unknown context captures full user-agent");
+    }
+
+    /**
+     * @dataProvider ipAddressInfo
+     */
+    public function canDetectUserIpAddress(UnitTester $I, Example $data)
+    {
+        $capture = new MockDirectEventCapture();
+
+        $client = new MockSentryClient($capture);
+
+        $request = new ServerRequest(
+            "GET",
+            "https://example.com/hello",
+            $data['headers']
+        );
+
+        $client->captureException(new RuntimeException("boom"), $request);
+
+        $body = json_decode($capture->requests[0]->body, true);
+
+        $I->assertEquals(
+            $data['expected'],
+            $body["user"]["ip_address"],
+            "{$data['why']} - headers: " . json_encode($data['headers'])
+        );
+    }
+
+    public function canCaptureBreadcrumbsViaPsr3LoggerAdapter(UnitTester $I)
+    {
+        $logger = new MockBreadcrumbLogger();
+        $capture = new MockDirectEventCapture();
+        $client = new MockSentryClient($capture, [$logger]);
+
+        $logger->info("hello world", ["foo" => "bar"]);
+        $logger->warning("Danger, Mr. Robinson!");
+
+        $request = new ServerRequest(
+            "GET",
+            "https://example.com/hello"
+        );
+
+        $client->captureException(new RuntimeException("boom"), $request);
+
+        $body = json_decode($capture->requests[0]->body, true);
+
+        $I->assertEquals(
+            [
+                [
+                    "timestamp" => $logger->time,
+                    "level"     => Level::INFO,
+                    "message"   => "[info] hello world",
+                    "data"      => ["foo" => "bar"],
+                ],
+                [
+                    "timestamp" => $logger->time,
+                    "level"     => Level::WARNING,
+                    "message"   => "[warning] Danger, Mr. Robinson!",
+                ],
+            ],
+            $body["breadcrumbs"]["values"],
+            "Can delegate log events to breadcrumbs"
+        );
+    }
+
+    public function canBufferAndManuallyFlushEvents(UnitTester $I)
+    {
+        $destination = new MockDirectEventCapture();
+        $buffer = new BufferedEventCapture($destination);
+        $client = new MockSentryClient($buffer);
+
+        $client->captureException($I->createExceptionWith("boom"));
+        $client->captureException($I->createExceptionWith("ouch"));
+
+        $I->assertCount(0, $destination->requests);
+
+        $buffer->flush();
+
+        $I->assertCount(2, $destination->requests, "Can flush buffered Events");
+
+        $destination->requests = [];
+
+        $buffer->flush();
+
+        $I->assertCount(0, $destination->requests, "Flushed events get cleared from the buffer");
+    }
+
+    protected function captureValues(): array
+    {
+        return [
+            ['expected' => 'array[3]', 'value' => [1, 2, 3]],
```

Comment:
Really nice job porting the tests! :+1:

Maybe put the indices for the dataprovider arrays in private consts, and maybe split the arrays into multiple lines. But that's small stuff :-)